torch/cuda/amp/amp_scaler.py

```diff
@@ -0,0 +1,394 @@
+
+        Simple example, using :meth:`unscale` to enable clipping of unscaled gradients::
+
+            ...
+            scaler.scale(loss).backward()
+            scaler.unscale(optimizer)
+            torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm)
+            scaler.step(optimizer)
+            scaler.update()
+
+        If this instance of :class:`AmpScaler` is not enabled, :meth:`unscale` is a no-op.
+
+        Arguments:
+            optimizer (torch.optim.Optimizer):  Optimizer that owns the gradients to be unscaled.
+
+        .. note::
+            :meth:`unscale` does not incur a CPU-GPU sync.
+
+        .. warning::
+            :meth:`unscale` should only be called once per optimizer per step,
+            and only after all gradients for that optimizer's owned parameters
+            have been accumulated.
+        """
+        if not self._enabled:
+            return
+
+        assert self._scale is not None, self._scale_not_initialized_error("unscale")
+
+        optimizer_state = self._per_optimizer_states[id(optimizer)]
+
+        if "unscaled" in optimizer_state:
+            raise RuntimeError("unscale() has already been called on this optimizer this iteration.")
+
+        # FP32 division can be imprecise for certain compile options, so we carry out the reciprocal in FP64.
+        inv_scale = self._scale.double().reciprocal().float()
+        found_inf = torch.full((1,), 0.0, dtype=torch.float32, device=self._scale.device)
+
+        optimizer_state["found_inf_per_device"] = self._unscale_grads(optimizer, inv_scale, found_inf)
+        optimizer_state["unscaled"] = True
+
+    def step(self, optimizer, *args, **kwargs):
+        """
+        Carry out a scaling-safe ``optimizer.step()``.  "Scaling-safe" means two things:
+
+        1.  , :meth:`step` invokes :meth:`unscale` for ``optimizer`` before calling ``optimizer.step()``
+            (unless :meth:`unscale` was explicitly called for ``optimizer`` earlier in the iteration).
+            This ensures ``optimizer.step()`` is carried out using unscaled gradients.
+        2.  If inf/NaN gradients are found, :meth:`step` skips ``optimizer.step()`` to avoid polluting the params.
+
+        ``*args`` and ``**kwargs`` are forwarded to ``optimizer.step()``.
+
+        Arguments:
+            optimizer (torch.optim.Optimizer):  Optimizer that applies the gradients.
+            args:  Any arguments.
+            kwargs:  Any keyword arguments.
+
+        Returns:
+            The return value of ``optimizer.step(*args, **kwargs)``.
+
+        .. warning::
+            Closure use is not currently supported.
+        """
+        if (not self._enabled):
+            return optimizer.step(*args, **kwargs)
+
+        if "closure" in kwargs:
+            raise RuntimeError("Closure use is not currently supported if AmpScaler is enabled.")
+
+        assert self._scale is not None, self._scale_not_initialized_error("step")
+
```

Comment:
Greg: I am concerned about multiple optimizers where some step and others don't.

We're not sure there's a nice automatic API to unscale multiple optimizers in advance to check if any would not step and then not step all of them. But it'd be nice if there was.

Reply:
There is, we talked about it briefly, and dismissed it as not worth the complexity for the first version.

AmpScaler constructor could additionally accept an instance of each optimizer that you might every use.  During update() at the end of each iteration, it would check if any of those optimizers were not stepped this iteration.  If any optimizer was not stepped _and_ still owned gradients, update() would adjust these gradient to be scaled by the scale factor that update() is setting for the next iteration, so that the next iteration's gradient scaled gradients would accumulate nicely on top of that optimizer's already-present gradients.